mlflow/gateway/client.py

```diff
@@ -112,6 +117,114 @@ def search_routes(self, search_filter: Optional[str] = None):
+            model configurations and how to dynamically create new routes within Databricks.
+
+
+        Example usage from within Databricks:
+
+        .. code-block:: python
+
+            from mlflow.gateway import MlflowGatewayClient
+
+            gateway_client = MlflowGatewayClient("databricks")
+
+            openai_api_key = ...
+
+            new_route = gateway_client.create_route(
+                "my-new-route",
+                "llm/v1/completions",
+                {
+                    "name": "question-answering-bot-1",
+                    "provider": "openai",
+                    "config": {
+                        "openai_api_key": openai_api_key,
+                        "openai_api_version": "2023-05-10",
+                        "openai_api_type": "openai/v1/chat/completions",
+                    },
+                },
+            )
+
+        """
+        if not self._is_databricks_host():
+            raise MlflowException(
+                "The create_route API is only available when running within "
+                "Databricks. Route creation is handled through creating a "
+                "configuration YAML file during startup or through updating a "
+                "running Gateway server.",
+                error_code=BAD_REQUEST,
+            )
+        payload = {
+            "name": name,
+            "route_type": route_type,
+            "model": model,
+        }
+        response = self._call_endpoint("POST", self._route_base, json.dumps(payload)).json()
+        return Route(**response)
+
+    def delete_route(self, name: str) -> None:
+        """
+        Delete an existing route in the Gateway.
+
+        .. warning::
+
+            This API is **only available** when running within Databricks. When running elsewhere,
+            route deletion is handled by removing the corresponding entry from the route
+            configuration YAML file that is specified during Gateway server start.
+
+        :param name: The name of the route to delete.
+
+        :raises mlflow.MlflowException: If the function is not running within Databricks.
+
+        Example usage from within Databricks:
+
+        .. code-block:: python
+
+            from mlflow.gateway import MlflowGatewayClient
+
+            gateway_client = MlflowGatewayClient("databricks")
+            gateway_client.delete_route("my-existing-route")
+
+        """
+        if not self._is_databricks_host():
+            raise MlflowException(
+                "The delete_route API is only available when running within Databricks. Route "
+                "deletion is handled through uploading a modified configuration YAML file to the "
+                "location specified when starting the Gateway server. To delete a route, remove "
+                "the route entry from the configuration file.",
+                error_code=BAD_REQUEST,
+            )
+        route = urljoin(self._route_base, name)
```

Comment:
I think our `urljoin` utility has some sensitivity to leading / trailing slashes. For example:

```
        route_name = "myroute"
        route = urljoin("/gateway", route_name)
        print("ROUTE 1", route)
        query_route = urljoin(route, "/invocations")
        print("ROUTE 2", query_route)
```

prints

```
ROUTE 1 /myroute
ROUTE 2 /invocations
```

but 

```
        route_name = "myroute"
        route = urljoin("/gateway/", route_name + "/")
        print("ROUTE 1", route)
        query_route = urljoin(route, "invocations")
        print("ROUTE 2", query_route)
```

prints 

```
ROUTE 1 /gateway/myroute/
ROUTE 2 /gateway/myroute/invocations
```

which is correct

Reply:
I think I'm going to just stop using that urllib.parse function and just create a utility. I'll push a fix momentarily :)